src/node_crypto.cc

```diff
@@ -3206,21 +3291,31 @@ bool CipherBase::Update(const char* data,
     return 0;
 
   // on first update:
-  if (kind_ == kDecipher && IsAuthenticatedMode() && auth_tag_len_ > 0) {
+  if (kind_ == kDecipher && IsAuthenticatedMode() && auth_tag_len_ > 0 &&
+      !auth_tag_set_) {
     EVP_CIPHER_CTX_ctrl(ctx_,
                         EVP_CTRL_GCM_SET_TAG,
                         auth_tag_len_,
                         reinterpret_cast<unsigned char*>(auth_tag_));
-    auth_tag_len_ = 0;
+    auth_tag_set_ = true;
   }
 
   *out_len = len + EVP_CIPHER_CTX_block_size(ctx_);
   *out = Malloc<unsigned char>(static_cast<size_t>(*out_len));
-  return EVP_CipherUpdate(ctx_,
-                          *out,
-                          out_len,
-                          reinterpret_cast<const unsigned char*>(data),
-                          len);
+  int r = EVP_CipherUpdate(ctx_,
+                           *out,
+                           out_len,
+                           reinterpret_cast<const unsigned char*>(data),
+                           len);
+
+  // When in CCM mode, EVP_CipherUpdate will fail if the authentication tag is
+  // invalid. In that case, remember the error and throw in final().
+  const int mode = EVP_CIPHER_CTX_mode(ctx_);
+  if (!r && kind_ == kDecipher && mode == EVP_CIPH_CCM_MODE) {
+    pending_auth_failed_ = true;
+    return true;
```

Comment:
so, what if the user doesn't call `final()`? Is there a scenario where they program their interface wrong and assume that a successful `update()` means all is good or is will they get caught out in the end? I know you have that note in the docs but you know what programmers are like with docs ..

Reply:
We *could* make `update()` throw if authentication fails, but that would make it inconsistent with GCM, and if we ever decide to add more AEAD algorithms, they would most likely not be compatible with that design. Note that ´update()` won't return any plaintext data but an empty buffer if authentication fails no matter whether the user calls `final()` or not, so at least it is impossible to process the decrypted message. This is something OpenSSL is a bit inconsistent about and I am not sure which design I prefer, some other crypto libraries allow decryption without the correct tag.